File: Coding Problems - Python/Arrays/32 Longest Valid Parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        max_len = 0
        st = [-1]

        for i in range(len(s)):
            if s[i] == "(":
                st.append(i)
            else:
                st.pop()
                if not st:
                    st.append(i)
                max_len = max(max_len, i - st[-1]) 
        return max_len

File: Coding_Problems_Java/32 Longest Valid Parentheses.java
class Solution {
    public int longestValidParentheses(String s) {
        int max_len = 0;
        ArrayList<Integer> st = new ArrayList<>();
        st.add(-1);

        for(int i = 0; i < s.length(); i++){
            char ch = s.charAt(i);
            if (ch == '('){
                st.add(i);
            }
            else{
                st.remove(st.size()-1);
                if(st.isEmpty()){
                    st.add(i);
                }
                max_len = Math.max(max_len, i - st.get(st.size()-1));
            }
        }
        return max_len;
    }
}
